docs/contributing/Nullable Annotations.md

```diff
@@ -0,0 +1,66 @@
+# Nullable Annotations
+
+This document describes how nullable annotations should be approached in the 
+Roslyn code base. The default is to simply follow [the same guidance](https://github.com/dotnet/runtime/blob/master/docs/coding-guidelines/api-guidelines/nullability.md)
+as the [dotnet/runtime](github.com/dotnet/runtime) repository. This document
+serves to detail the places where the guidance differs for Roslyn and 
+re-emphasize rules that come up frequently.
+
+## Annotation Guidance 
+
+- **DO** annotate all new code
+- **DO** use the null suppression operator in places where a method body has
+ensured a value cannot be `null` but the compiler is unable to track it. 
+Example:
+```cs
+struct SyntaxTrivia {
+    void Example() {    
+        if (this.HasStructure) {
+            Use(this.GetStructure()!)
+        }
+    }
+}
+```
+- **DO** use `Assert` to validate state that is passed into the method
+```cs
+void M1(SyntaxNode node) {
+    RoslynDebug.Assert(node.Parent is object);
+    ...
+    M2(node.Parent);
+}
+```
+- **AVOID** annotating and refactoring code in the same commit. This can
+significantly increase the cost of reviewing. Prefer instead to separate 
+annotations and significant refactorings into separate commits. Small changes,
+and modernizing style (change `default(SyntaxToken)` to `default`) are
+acceptable.
```

Comment:
One example is renaming a parameter to remove the `Opt` suffix. I found it's often cleaner to avoid the rename initially, and then rename it as a future commit.

Reply:
The `Opt` suffix is borderline for me. Prefer we do it as a separate commit but not so much that I put up a fight over it.

---
In reply to: [366078764](https://github.com/dotnet/roslyn/pull/40940#discussion_r366078764) [](ancestors = 366078764)

Reply:
The one thing I found icky about doing Opt in a separate commit was having to go back and find all the places you have to fix up; it's easy to fix both the annotation and name at once.